File: Pokedex/Pokedex/Models/Pokemon.swift
//
//  Pokemon.swift
//  Pokedex
//
//  Created by Thomas Bernard on 01/07/2023.
//

import Foundation

struct Pokemon : Decodable, Identifiable {
    
    let id : Int
    let name : String
    let types : [Type]
    let description : String
    let height : Int
    let weight : Float
    let sprites : PokemonSprites
    let stats : [PokemonStat]
    let abilities : [String]
    let growthRate : String
    let captureRate : Int
    let baseHappiness : Int
    let shape : String
    let evolutionChain : PokemonEvolutionChain
    
    
    enum CodingKeys: String, CodingKey {
        case id
        case name
        case types
        case description
        case height
        case weight
        case sprites
        case stats
        case abilities
        case shape
        case growthRate = "growth_rate"
        case captureRate = "capture_rate"
        case baseHappiness = "base_happiness"
        case evolutionChain = "evolution_chain"
    }
}

File: PokemonGrabber/PokemonGrabber/Pokemon.cs
using Newtonsoft.Json;

namespace PokemonGrabber;

public class Pokemon
{
    [JsonProperty("id")]
    public int Id { get; set; }
    
    [JsonProperty("name")]
    public string Name { get; set; }
    
    [JsonProperty("weight")]
    public float Weight { get; set; }
    
    [JsonProperty("height")]
    public int Height { get; set; }
    
    [JsonProperty("description")]
    public string Description { get; set; }
    
    [JsonProperty("growth_rate")]
    public string GrowthRate { get; set; }
    
    [JsonProperty("base_happiness")]
    public int BaseHappiness { get; set; } 
    
    [JsonProperty("shape")]
    public string Shape { get; set; } 
    
    [JsonProperty("capture_rate")]
    public int CaptureRate { get; set; }
    
    [JsonProperty("types")]
    public List<string> Types { get; set; }
    
    [JsonProperty("sprites")]
    public PokemonSprite Sprites { get; set; }
    
    [JsonProperty("stats")]
    public IEnumerable<PokemonStat> Stats { get; set; }
    
    [JsonProperty("abilities")]
    public IEnumerable<string> Abilities { get; set; }
    
    [JsonProperty("evolution_chain")]
    public PokemonEvolutionChain? EvolutionChain { get; set; }
}

public class PokemonStat
{
    [JsonProperty("name")]
    public string Name { get; set; }
    
    [JsonProperty("value")]
    public int Value { get; set; }
}

public class PokemonEvolutionChain
{
    [JsonProperty("pokemon_id")]
    public int PokemonId { get; set; }
    
    [JsonProperty("level")]
    public int? Level { get; set; }
    
    [JsonProperty("item")]
    public string? Item { get; set; }
    
    [JsonProperty("evolution_chain")]
    public IEnumerable<PokemonEvolutionChain>? EvolutionChain { get; set; }
}

public class PokemonSprite
{
    [JsonProperty("front_default_official_artwork")]
    public string FrontDefaultOfficialArtwork { get; set; }
    
    [JsonProperty("front_shiny_official_artwork")]
    public string FrontShinyOfficialArtwork { get; set; }
    
    //[JsonProperty("front_male_dream_world")]
    //public string FrontMaleDreamWorld { get; set; }
    //
    //[JsonProperty("front_female_dream_world")]
    //public string FrontFemaleDreamWorld { get; set; }
}